modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/qa/matchers/EqualMatcher.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.datastreams.logsdb.qa.matchers;
+
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.datastreams.logsdb.qa.exceptions.MatcherException;
+import org.elasticsearch.datastreams.logsdb.qa.exceptions.MismatchTypeMatcherException;
+import org.elasticsearch.datastreams.logsdb.qa.exceptions.UncomparableMatcherException;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.util.Arrays;
+import java.util.List;
+
+class EqualMatcher<T> extends Matcher {
+    private final XContentBuilder actualMappings;
+    private final Settings.Builder actualSettings;
+    private final XContentBuilder expectedMappings;
+    private final Settings.Builder expectedSettings;
+    private final T actual;
+    private final T expected;
+    private final boolean ignoringSort;
+
+    EqualMatcher(
+        XContentBuilder actualMappings,
+        Settings.Builder actualSettings,
+        XContentBuilder expectedMappings,
+        Settings.Builder expectedSettings,
+        T actual,
+        T expected,
+        boolean ignoringSort
+    ) {
+        this.actualMappings = actualMappings;
+        this.actualSettings = actualSettings;
+        this.expectedMappings = expectedMappings;
+        this.expectedSettings = expectedSettings;
+        this.actual = actual;
+        this.expected = expected;
+        this.ignoringSort = ignoringSort;
+    }
+
+    @SuppressWarnings("unchecked")
+    public boolean match() throws MatcherException {
+        if (actual == null) {
+            if (expected == null) {
+                throw new UncomparableMatcherException(
+                    actualMappings,
+                    actualSettings,
+                    expectedMappings,
+                    expectedSettings,
+                    "Both 'actual' and 'expected' are null"
+                );
+            }
+            return false;
+        }
+        if (expected == null) {
+            return false;
+        }
+        if (actual.getClass().equals(expected.getClass()) == false) {
+            throw new MismatchTypeMatcherException(
+                actualMappings,
+                actualSettings,
+                expectedMappings,
+                expectedSettings,
+                "Unable to match " + actual.getClass().getSimpleName() + " to " + expected.getClass().getSimpleName()
+            );
+        }
+        if (actual.getClass().isArray()) {
+            return matchArraysEqual((T[]) actual, (T[]) expected, ignoringSort);
+        }
+        if (actual instanceof List<?> && expected instanceof List<?>) {
```

Comment:
nit: you can get rid of one cast if you bind here `actual instanceof List<?> actualList`